domain/port/state/state.go

```diff
@@ -211,86 +190,164 @@ AND endpoint IN ($endpoints[:])
+			for _, openedPortRange := range currentOpenedPorts {
+				if portRange == openedPortRange.decode() && ep.Endpoint == openedPortRange.Endpoint {
+					found = true
+					break
+				}
+			}
+			if found {
+				continue
+			}
+
+			uuid, err := uuid.NewUUID()
+			if err != nil {
+				return errors.Annotatef(err, "generating UUID for unit endpoint")
+			}
+			openPortRanges = append(openPortRanges, unitPortRange{
+				UUID:             uuid.String(),
+				ProtocolID:       protocolMap[portRange.Protocol],
+				FromPort:         portRange.FromPort,
+				ToPort:           portRange.ToPort,
+				UnitEndpointUUID: ep.UUID,
+			})
+		}
 	}
 
-	return endpointToPorts, nil
+	if len(openPortRanges) > 0 {
+		err = tx.Query(ctx, insertPortRange, openPortRanges).Run()
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
 }
 
-// updateEndpointOpenedPorts updates the opened ports for the given endpoints.
-func (st *State) updateEndpointOpenedPorts(
-	ctx context.Context, tx *sqlair.TX, endpointUUIDs endpointUUIDs, unitPortRanges []unitPortRange,
-) error {
-	insertPortRange, err := st.Prepare("INSERT INTO port_range (*) VALUES ($unitPortRange.*)", unitPortRange{})
+// getProtocolMap returns a map of protocol names to their IDs in DQLite.
+func (st *State) getProtocolMap(ctx context.Context, tx *sqlair.TX) (map[string]int, error) {
+	getProtocols, err := st.Prepare("SELECT &protocol.* FROM protocol", protocol{})
 	if err != nil {
-		return errors.Annotate(err, "preparing insert port range statement")
+		return nil, errors.Annotate(err, "preparing get protocol ID statement")
+	}
+
+	protocols := []protocol{}
+	err = tx.Query(ctx, getProtocols).GetAll(&protocols)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	protocolMap := map[string]int{}
+	for _, protocol := range protocols {
+		protocolMap[protocol.Name] = protocol.ID
 	}
 
-	clearPortRange, err := st.Prepare(`
+	return protocolMap, nil
+}
+
+// closePorts removes the given port ranges from the database, if they exist.
+func (st *State) closePorts(
+	ctx context.Context, tx *sqlair.TX, closePorts network.GroupedPortRanges, currentOpenedPorts []endpointPortRangeUUID,
+) error {
+	closePortRanges, err := st.Prepare(`
 DELETE FROM port_range
-WHERE unit_endpoint_uuid IN ($endpointUUIDs[:])
-`, endpointUUIDs)
+WHERE uuid IN ($portRangeUUIDs[:])
+`, portRangeUUIDs{})
 	if err != nil {
-		return errors.Annotate(err, "preparing clear port range statement")
+		return errors.Annotate(err, "preparing close port range statement")
 	}
 
-	// Clear out all existing open ports for endpoints undergoing changes.
-	if len(endpointUUIDs) > 0 {
-		err := tx.Query(ctx, clearPortRange, endpointUUIDs).Run()
-		if err != nil {
-			return errors.Trace(err)
+	// Find the uuids of port ranges to close
+	var closePortRangeUUIDs portRangeUUIDs
+	for endpoint, portRanges := range closePorts {
+		for _, closePortRange := range portRanges {
+			for _, openedPortRange := range currentOpenedPorts {
```

Comment:
Same goes here. Pre-calculate as much as possible and then reduce the number of inner loops.

Reply:
Same here. Turns out in this case I can also construct an index. Not sure how I missed that :sweat_smile: